### PR TITLE
FINERACT-2017: Fix apply holidays to loans job

### PR DESCRIPTION
**Pull request by mariiaKraievska**
## Description

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket]
JIRA ticket - https://issues.apache.org/jira/browse/FINERACT-2017

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [x] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [x] Create/update unit or integration tests for verifying the changes made.

- [x] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [x] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [x] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/SchedulerJobsTestResults.java [hunk 1]
```diff
@@ -329,21 +331,49 @@ public void testApplyHolidaysToLoansJobOutcome() throws InterruptedException {
...
+
+        for (LinkedHashMap period : periods) {
+            ArrayList<Integer> fromDate = (ArrayList<Integer>) period.get("fromDate");
+            if (fromDate != null && Objects.equals(fromDate.get(1), repaymentsRescheduledDate.get(1))) {
+                dateToApplyHolidays = fromDate;
+            }
+        }
 
-        Assertions.assertEquals(repaymentsRescheduledDate, repaymentsRescheduledDate,
-                "Verifying Repayment Rescheduled Date after Running Apply Holidays to Loans Scheduler Job");
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/SchedulerJobsTestResults.java:346]: This test is not really testing whether only the schedule that fall on that particular date was rescheduled. You should check whether the following periods after the one that got rescheduled was not moved.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/jobs/applyholidaystoloans/ApplyHolidaysToLoansTasklet.java [hunk 2]
```diff
@@ -129,13 +131,13 @@ private void adjustRepaymentSchedules(Loan loan, Holiday holiday, LocalDate adju
...
-            if (!DateUtils.isBefore(oldDueDate, holiday.getFromDate())) {
+            if (isDateWithinRange(oldDueDate, holiday.getFromDate(), holiday.getToDate())) {
                 // FIXME: AA do we need to apply non-working days.
                 // Assuming holiday's repayment reschedule to date cannot be
                 // created on a non-working day.
 
-                adjustedRescheduleToDate = scheduledDateGenerator.generateNextRepaymentDate(adjustedRescheduleToDate, loanApplicationTerms,
-                        false);
+                adjustedRescheduleToDate = scheduledDateGenerator.generateNextRepaymentDateWhenHolidayApply(adjustedRescheduleToDate,
+                        loanApplicationTerms);
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/jobs/applyholidaystoloans/ApplyHolidaysToLoansTasklet.java:140]: Can you please fix an another issue?
Line 90: holiday.isProcessed() -> holiday.setProcessed(true);

**Comment by adamsaghy** [comment]
I am merging it and raise a separate fix for processing holidays.

**Comment by mariiaKraievska** [comment]
> I am merging it and raise a separate fix for processing holidays.

Thank you. I will take care of it. Could you please share the jira tiket number for me?

**Comment by adamsaghy** [comment]
@mariiaKraievska https://github.com/apache/fineract/pull/3647 I have raised the fix in this PR.

I was not sure whether you are available before the holidays or not, so that's why I decided to merge your PR and raise an another one for this.

**Comment by adamsaghy** [comment]
@mariiaKraievska For improving the test:
"This test is not really testing whether only the schedule that fall on that particular date was rescheduled. You should check whether the following periods after the one that got rescheduled was not moved."

Feel free to raise an another PR for the same story. It is not crucial anymore.

**Comment by mariiaKraievska** [comment]
> @mariiaKraievska For improving the test: "This test is not really testing whether only the schedule that fall on that particular date was rescheduled. You should check whether the following periods after the one that got rescheduled was not moved."
> 
> Feel free to raise an another PR for the same story. It is not crucial anymore.

Sure, I will do this. Thank you. I'm available, so if there are any other tasks that need to be done, feel free to contact me, I'll be happy to do it.

**Comment by edcable** [comment]
@mariiaKraievska thanks for all the contributions you've been making over the past several months - it's gladly appreciated. what is best means of contacting you as I do have some other high priority tasks that could greatly benefit the community I'd like to discuss.

**Comment by mariiaKraievska** [comment]
> @mariiaKraievska thanks for all the contributions you've been making over the past several months - it's gladly appreciated. what is best means of contacting you as I do have some other high priority tasks that could greatly benefit the community I'd like to discuss.

To be honest, I am a member of the [Onix-Systems](https://onix-systems.com/) team, and we are deeply interested in harnessing the potential of Fineract. We want to contribute more actively to the project and involve more of our developers. Please contact my CTO, Denis Sheremetov - [denis@onix-systems.com](mailto:denis@onix-systems.com)

**Comment by mrjazz** [comment]
Hi @edcable! Thank you for noting our contribution. Fineract is a very interesting and mature product, and we are glad that we were able to put our efforts into its development. In the future, we would like to be more valuable contributors to the project and work on more complex and important tasks. If you don't mind, let's schedule a call and discuss how we could be useful to the project and the project to us. You can write me on email.

**Pull request merged**